kernel/shared/src/main/scala/monocle/optics.scala

```diff
@@ -0,0 +1,197 @@
+    new PLens[S, T, A, B] {
+      def get(s: S): A       = get0(s)
+      def set(to: B): S => T = set0(_, to)
+    }
+}
+
+sealed trait POptional[-S, +T, +A, -B] extends PSetter[S, T, A, B] { self =>
+
+  def getOrModify(from: S): Either[T, A]
+
+  def set(to: B): S => T
+
+  final override def fold[Z](zero: Z, combine: (Z, Z) => Z)(f: A => Z)(s: S): Z =
+    getOrModify(s).fold(_ => zero, a => f(a))
+
+  final def getOption(from: S): Option[A] =
+    getOrModify(from).toOption
+
+  final def modify(f: A => B): S => T =
+    s => getOrModify(s).fold(t => t, a => set(f(a))(s))
+
+  final def compose[C, D](other: POptional[A, B, C, D]): POptional[S, T, C, D] =
+    new POptional[S, T, C, D] {
+      def getOrModify(from: S): Either[T, C] =
+        self.getOrModify(from).flatMap(a => other.getOrModify(a).left.map(self.set(_)(from)))
+      def set(to: D): S => T =
+        self.modify(other.set(to))
+    }
+}
+
+object POptional {
+  def apply[S, T, A, B](getOrModify0: S => Either[T, A])(set0: (S, B) => T): POptional[S, T, A, B] =
+    new POptional[S, T, A, B] {
+      def getOrModify(from: S): Either[T, A] = getOrModify0(from)
+      def set(to: B): S => T                 = set0(_, to)
+    }
+}
+
+sealed trait PPrism[-S, +T, +A, -B] extends POptional[S, T, A, B] { self =>
+
+  def reverseGet(to: B): T
+
+  final def set(to: B): S => T =
+    _ => reverseGet(to)
+
+  final def compose[C, D](other: PPrism[A, B, C, D]): PPrism[S, T, C, D] =
+    new PPrism[S, T, C, D] {
+      def getOrModify(from: S): Either[T, C] =
+        self.getOrModify(from).flatMap(a => other.getOrModify(a).left.map(self.set(_)(from)))
+      def reverseGet(to: D): T =
+        self.reverseGet(other.reverseGet(to))
+    }
+}
+
+object PPrism {
+  def apply[S, T, A, B](getOption0: S => Either[T, A])(reverseGet: B => T): PPrism[S, T, A, B] =
+    new PPrism[S, T, A, B] {
+      def getOrModify(from: S): Either[T, A] = getOption0(from)
+      def reverseGet(to: B): T               = reverseGet(to)
+    }
+}
+
+sealed trait PSetter[-S, +T, +A, -B] extends Fold[S, A] { self =>
+
+  def fold[Z](zero: Z, combine: (Z, Z) => Z)(f: A => Z)(s: S): Z =
+    zero
+
+  /** modify polymorphically the target of a [[PSetter]] with a function */
+  def modify(f: A => B): S => T
+
+  /** set polymorphically the target of a [[PSetter]] with a value */
+  def set(b: B): S => T
+}
+
+object PSetter {
+
+  def apply[S, T, A, B](modify0: (S, A => B) => T)(set0: (S, B) => T): PSetter[S, T, A, B] =
+    new PSetter[S, T, A, B] {
+      def modify(f: A => B): S => T = modify0(_, f)
+      def set(b: B): S => T = set0(_, b)
+
+    }
+
+}
+
+sealed trait PTraversal[S, T, A, B] extends PSetter[S, T, A, B] { self =>
+  def traverse[Z](z: Z, f: (Z, A) => Either[Z, (Z, B)]): S => (Z, T)
+
+  final override def fold[Z](zero: Z, combine: (Z, Z) => Z)(f: A => Z)(s: S): Z =
+    traverse[Z](zero, (z, a) => Left(combine(z, f(a))))(s)._1
```

Comment:
I started to play a bit with this encoding. I am not entirely sure what does it mean to return a Left or a Right in traverse. At first, I thought Left was a way to short cut the traversal but it doesn't look to be the case.

@jdegoes @adamgfraser do you have an example of a `Traversal` for all elements in a List.